File: Trie/src/Trie.java
import java.util.TreeMap;

public class Trie {   // 多叉树

    private class Node{

        public boolean isWord;   // 判断是否是一个单词
        public TreeMap<Character, Node> next;   // 下一节点是这个字符对引得一个映射

        public Node(boolean isWord){
            this.isWord = isWord;
            next = new TreeMap<>();
        }

        public Node(){
            this(false);
        }
    }

    private Node root;
    private int size;  // 存储单词数

    public Trie(){
        root = new Node();
        size = 0;
    }

    // 获得trie中的单词数
    public int getSize(){
        return size;
    }

    // 向trie中添加一个单词
    public void add(String word){
        Node cur = root;
        for (int i = 0; i < word.length(); i ++){   // 找到下一个节点的映射中是否存在该字符对应的映射
            char c = word.charAt(i);
            if (cur.next.get(c) == null){   // 如果这个字符的下一个节点为空
                cur.next.put(c, new Node());
            }
            cur = cur.next.get(c);
        }

        if (! cur.isWord){   // 如果这个单词之前没有
            cur.isWord = true;
            size ++;
        }
    }

    // 在trie中查询单词
    public boolean contains(String word){

        Node cur = root;

        for (int i = 0 ; i < word.length(); i ++){
            char c = word.charAt(i);
            if (cur.next.get(c) == null){
                return false;
            }
            cur = cur.next.get(c);
        }

        return cur.isWord;
    }


    // 查找是否在trie中有单词以prefix为前缀
    public boolean isPrefix(String prefix){

        Node cur = root;
        for (int i = 0; i < prefix.length(); i ++){
            char c = prefix.charAt(i);
            if (cur.next.get(c) == null){
                return false;
            }
            cur = cur.next.get(c);
        }

        return true;
    }
}

File: _implement/Trie.py


class TrieNode:
    def __init__(self, isWord=None):
        if isWord == None:
            self.isWord = False
        else:
            self.isWord = isWord

        self.Node = TrieNode
        self.next = {}   # key: 字符， value：下一个节点


class Trie:
    def __init__(self):
        self.__root = self.__getNode()

    def __getNode(self):
        return TrieNode()


    # 非递归实现trie添加元素操作
    def add(self, word):
        cur = self.__root
        for i in range(len(word)):
            c = word[i]
            if cur.next.get(c) == None:
                cur.next[c] = self.__getNode()
            cur = cur.next.get(c)

        if cur.isWord == False:
            cur.isWord = True

    # 使用循环进行trie添加元素操作
    def add_recursion(self, word):
        cur = self.__root
        self.__addR(cur, word, 1)

    def __addR(self, TrieNode, word, index):
        c = word[index - 1]
        if index == len(word):
            TrieNode.isWord = True
            return
        if TrieNode.next.get(c) == None:
            TrieNode.next[c] = self.__getNode()

        self.__addR(TrieNode.next.get(c), word, index + 1)

    # 在trie中查询单词
    def contains(self, word):
        cur = self.__root

        for i in range(len(word)):
            c = word[i]
            if cur.next.get(c) == None:
                return False
            cur = cur.next.get(c)

        return cur.isWord


    # 是否有以word为前缀的单词
    def isPrefix(self, word):
        cur = self.__root

        for i in range(len(word)):
            c = word[i]

            if cur.next.get(c) == None:
                return False
            cur = cur.next.get(c)

        return True

    # 打印整个trie树，像git log中打印出的表相同
    def PrintTrie(self):
        pass

if __name__ == "__main__":
    Trie = Trie()
    Trie.add("worgggggggggggggggggd")
    Trie.add("Trie")
    Trie.add("Tree")
    Trie.add("Train")
    Trie.add("fish")
    print(Trie.contains("Trie"))
